File: win_app_2/src/services/auth/authenticationManager.ts
import { createLogger } from '@/lib/logger'
import { captureError, addBreadcrumb, setUser as setSentryUser, clearUser as clearSentryUser } from '@/services/crash/crashReporter'
import { useAuthStore } from '@/stores/authStore'
import * as authApi from './authApiClient'
import { setOnSessionExpired } from './authApiClient'
import { startGoogleAuth } from './googleAuthService'
import * as licenseManager from '@/services/license/licenseManager'
import type { AuthUser, DeviceInfo } from '@/types/auth'

const log = createLogger('Auth')

// Wire mid-session expiration detection at module level
setOnSessionExpired(() => useAuthStore.getState().setSessionExpired())

let pollingInterval: ReturnType<typeof setInterval> | null = null

// Fallback device ID - persisted in memory for consistency across the session
let fallbackDeviceId: string | null = null

function getFallbackDeviceId(): string {
  if (!fallbackDeviceId) {
    fallbackDeviceId = crypto.randomUUID()
  }
  return fallbackDeviceId
}

async function getDeviceInfo(): Promise<DeviceInfo> {
  const info = await window.electronAPI?.getDeviceInfo()
  return (
    info || {
      deviceId: getFallbackDeviceId(),
      deviceName: 'Unknown',
      platform: 'windows' as const,
      osVersion: 'Unknown',
      appVersion: '1.0.0',
    }
  )
}

// Export for use by syncManager
export async function getDeviceId(): Promise<string> {
  const info = await getDeviceInfo()
  return info.deviceId
}

async function storeTokens(
  accessToken: string,
  refreshToken: string,
  expiresIn: number,
  user: AuthUser,
): Promise<void> {
  authApi.setAccessToken(accessToken, expiresIn)
  await window.electronAPI?.secureStore.set('refresh_token', refreshToken)
  await window.electronAPI?.store.set('auth.user', JSON.stringify(user))
}

export async function checkExistingAuth(): Promise<void> {
  const store = useAuthStore.getState()

  const userJson = (await window.electronAPI?.store.get('auth.user')) as string | null
  const hasRefreshToken = await window.electronAPI?.secureStore.has('refresh_token')

  if (!userJson || !hasRefreshToken) {
    log.info('No stored credentials found')
    store.setUnauthenticated()
    return
  }

  const user: AuthUser = JSON.parse(userJson)
  log.info(`Found stored credentials for: ${user.email}`)

  try {
    const refreshToken = await window.electronAPI?.secureStore.get('refresh_token')
    if (!refreshToken) {
      store.setUnauthenticated()
      return
    }

    const response = await authApi.refreshTokens(refreshToken)
    authApi.setAccessToken(response.accessToken, response.expiresIn)
    await window.electronAPI?.secureStore.set('refresh_token', response.refreshToken)

    store.setAuthenticated(user)
    setSentryUser(user.id, user.email)
    addBreadcrumb('auth', 'Authentication restored from stored credentials', 'info')
    log.info(`Authentication restored for: ${user.email}`)

    // Revalidate license after restoring auth
    licenseManager.revalidate().catch((err) => {
      log.warn('License revalidation failed:', err)
    })
  } catch (error) {
    const errorCode = error instanceof Error && 'code' in error ? (error as { code?: string }).code : ''
    const errorStr = String(error).toLowerCase()

    // Permanent errors: force full logout (matches macOS isPermanentAuthError)
    const isPermanent =
      errorCode === 'account_blocked' ||
      errorCode === 'device_limit' ||
      errorStr.includes('account_blocked') ||
      errorStr.includes('device_limit')

    if (isPermanent) {
      log.warn('Permanent auth error, forcing logout:', errorStr)
      captureError(
        error instanceof Error ? error : new Error('Permanent auth error'),
        { service: 'auth', errorCode, reason: 'permanent' },
      )
      await clearCredentials()
      store.setUnauthenticated()
      return
    }

    // Auth rejection: session expired
    const isAuthRejection =
      errorStr.includes('invalid_token') ||
      errorStr.includes('token_revoked') ||
      errorStr.includes('401') ||
      errorStr.includes('403')

    if (isAuthRejection) {
      log.warn('Server rejected credentials, session expired')
      await clearCredentials()
      store.setSessionExpired()
    } else {
      // Network/transient error - degraded mode: keep user authenticated with cached data
      // (matches macOS pattern: transparent retry on next action)
      log.warn(`[diag] DEGRADED MODE — auth check transient err: "${errorStr.substring(0, 200)}"`)
      log.warn('Transient error during auth check, entering degraded mode:', errorStr)
      store.setAuthenticated(user)
      setSentryUser(user.id, user.email)
    }
  }
}

export async function loginWithCredentials(email: string, password: string): Promise<void> {
  const store = useAuthStore.getState()
  store.setAuthenticating()

  try {
    const deviceInfo = await getDeviceInfo()
    const response = await authApi.login(email, password, deviceInfo)

    await storeTokens(response.accessToken, response.refreshToken, response.expiresIn, response.user)
    store.setAuthenticated(response.user)
    setSentryUser(response.user.id, response.user.email)
    addBreadcrumb('auth', 'User logged in with credentials', 'info')

    // Revalidate license after login
    licenseManager.revalidate().catch((err) => {
      log.warn('License revalidation failed:', err)
    })
  } catch (error) {
    store.setError(error instanceof Error ? error.message : 'Login failed')
    throw error
  }
}

export async function registerWithCredentials(
  name: string,
  email: string,
  password: string,
): Promise<void> {
  const store = useAuthStore.getState()
  store.setAuthenticating()

  try {
    const deviceInfo = await getDeviceInfo()
    const response = await authApi.register(name, email, password, deviceInfo)

    await storeTokens(response.accessToken, response.refreshToken, response.expiresIn, response.user)
    store.setAuthenticated(response.user)
    setSentryUser(response.user.id, response.user.email)
    addBreadcrumb('auth', 'User registered with credentials', 'info')

    // Revalidate license after registration
    licenseManager.revalidate().catch((err) => {
      log.warn('License revalidation failed:', err)
    })
  } catch (error) {
    store.setError(error instanceof Error ? error.message : 'Registration failed')
    throw error
  }
}

export async function startDeviceCodeFlow(): Promise<{
  userCode: string
  verificationUri: string
}> {
  const store = useAuthStore.getState()
  log.info('Starting device code flow...')
  store.setAuthenticating()

  try {
    const deviceInfo = await getDeviceInfo()
    log.info('Device info:', deviceInfo)

    log.info('Requesting device code from API...')
    const response = await authApi.requestDeviceCode(deviceInfo)
    log.info('Device code response:', {
      userCode: response.userCode,
      verificationUrl: response.verificationUrl,
      expiresIn: response.expiresIn,
    })

    const expiresAt = new Date(Date.now() + response.expiresIn * 1000).toISOString()
    store.setDeviceCodePending(response.userCode, response.deviceCode, expiresAt)
    log.info('Auth state set to deviceCodePending')

    // Start polling
    startPolling(response.deviceCode, response.expiresIn)
    log.info('Polling started')

    return {
      userCode: response.userCode,
      verificationUri: response.verificationUrl, // Keep return type consistent
    }
  } catch (error) {
    log.error('Device code flow failed:', error)
    captureError(
      error instanceof Error ? error : new Error('Device code flow failed'),
      { service: 'auth', method: 'device_code' },
    )
    store.setError(error instanceof Error ? error.message : 'Device code flow failed')
    throw error
  }
}

function startPolling(deviceCode: string, expiresIn: number): void {
  cancelDeviceCodeFlow()

  const expiresAt = Date.now() + expiresIn * 1000
  log.info(`Polling will expire at ${new Date(expiresAt).toISOString()}`)

  pollingInterval = setInterval(async () => {
    if (Date.now() > expiresAt) {
      log.warn('Device code expired')
      cancelDeviceCodeFlow()
      useAuthStore.getState().setError('Device code expired. Please try again.')
      return
    }

    try {
      log.info('Polling for device code authorization...')
      const response = await authApi.pollDeviceCode(deviceCode)
      log.info('Poll response:', { error: response.error, hasToken: !!response.accessToken })

      if (response.accessToken && response.refreshToken && response.user && response.expiresIn) {
        log.info('Authorization successful!')
        cancelDeviceCodeFlow()
        await storeTokens(
          response.accessToken,
          response.refreshToken,
          response.expiresIn,
          response.user,
        )
        useAuthStore.getState().setAuthenticated(response.user)
        setSentryUser(response.user.id, response.user.email)
        addBreadcrumb('auth', 'User authenticated via device code flow', 'info')

        // Revalidate license after device code auth
        licenseManager.revalidate().catch((err) => {
          log.warn('License revalidation failed:', err)
        })
        return
      }

      if (response.error && response.error !== 'authorization_pending') {
        log.warn('Poll returned error:', response.error)
        cancelDeviceCodeFlow()
        useAuthStore.getState().setError(response.message || response.error)
      }
    } catch (error) {
      // Ignore network errors during polling but log them
      log.warn('Polling error (will retry):', error)
    }
  }, 5000)
}

export function cancelDeviceCodeFlow(): void {
  if (pollingInterval) {
    clearInterval(pollingInterval)
    pollingInterval = null
  }
}

export async function loginWithGoogle(): Promise<void> {
  const store = useAuthStore.getState()
  log.info('Starting Google sign-in via OAuth flow')
  store.setAuthenticating()

  try {
    // Start OAuth flow - opens browser and waits for callback
    const result = await startGoogleAuth()

    if (!result.success || !result.code || !result.codeVerifier || !result.redirectUri) {
      log.error('Google OAuth failed:', result.error)
      // Stop OAuth server on failure
      await window.electronAPI?.auth?.stopOAuthServer()
      throw new Error(result.error || 'Google sign-in failed')
    }

    log.info('Google OAuth code received, exchanging with backend...')
    log.info('Backend URL:', result.redirectUri)

    // Exchange authorization code for tokens with our backend
    const deviceInfo = await getDeviceInfo()

    try {
      const response = await authApi.exchangeGoogleAuth(
        result.code,
        result.codeVerifier,
        result.redirectUri,
        deviceInfo,
      )

      log.info('Google sign-in successful')
      await storeTokens(response.accessToken, response.refreshToken, response.expiresIn, response.user)
      store.setAuthenticated(response.user)
      setSentryUser(response.user.id, response.user.email)
      addBreadcrumb('auth', 'User logged in via Google OAuth', 'info')

      // Revalidate license after Google sign-in
      licenseManager.revalidate().catch((err) => {
        log.warn('License revalidation failed:', err)
      })
    } finally {
      // Always stop OAuth server after token exchange attempt (success or failure)
      log.info('Stopping OAuth server after token exchange')
      await window.electronAPI?.auth?.stopOAuthServer()
    }
  } catch (error) {
    log.error('Google sign-in failed:', error)
    captureError(
      error instanceof Error ? error : new Error('Google sign-in failed'),
      { service: 'auth', method: 'google' },
    )
    store.setError(error instanceof Error ? error.message : 'Google sign-in failed')
    throw error
  }
}

export async function logoutUser(allDevices: boolean = false): Promise<void> {
  const store = useAuthStore.getState()
  cancelDeviceCodeFlow()

  try {
    const refreshToken = await window.electronAPI?.secureStore.get('refresh_token')
    await authApi.logout(refreshToken, allDevices)
  } catch {
    log.warn('Logout API call failed, continuing with local logout')
  }

  await clearCredentials()
  clearSentryUser()
  addBreadcrumb('auth', 'User logged out', 'info')
  store.logout()
}

async function clearCredentials(): Promise<void> {
  authApi.clearTokens()
  await window.electronAPI?.secureStore.delete('refresh_token')
  await window.electronAPI?.store.delete('auth.user')
}

export async function getAccessToken(force = false): Promise<string> {
  const token = await authApi.getValidAccessToken(force)
  if (!token) throw new Error('Not authenticated')
  return token
}

File: ios_app/QueenMamaIOS/Services/Auth/AuthenticationManager.swift
import Foundation
import Combine

/// Main authentication coordinator for the macOS app
/// Manages login flows, token refresh, and authentication state
@MainActor
final class AuthenticationManager: ObservableObject {
    static let shared = AuthenticationManager()

    // MARK: - Published State

    @Published private(set) var authState: AuthState = .unknown
    @Published private(set) var isAuthenticated: Bool = false
    @Published private(set) var currentUser: AuthUser?

    // MARK: - Dependencies

    private let api = AuthAPIClient.shared
    private let tokenStore = AuthTokenStore.shared
    private let deviceInfo = DeviceInfo.current()

    // MARK: - Device Code Polling

    private var pollingTask: Task<Void, Never>?
    private let pollingInterval: TimeInterval = 5

    private init() {}

    // MARK: - Initialization

    /// Check for existing authentication on app launch
    func checkExistingAuth() async {
        print("[Auth] ========== AUTH CHECK STARTED ==========")
        authState = .unknown

        // Check for stored credentials
        guard tokenStore.hasStoredCredentials,
              let storedUser = tokenStore.storedUser else {
            print("[Auth] No stored credentials found - user needs to login")
            print("[Auth] ==========================================")
            authState = .unauthenticated
            return
        }

        print("[Auth] Found stored credentials for user: \(storedUser.email.prefix(3))***")
        print("[Auth] Has refresh token: \(tokenStore.refreshToken != nil)")
        print("[Auth] Access token valid: \(tokenStore.isAccessTokenValid)")

        // Try to validate/refresh tokens
        do {
            if !tokenStore.isAccessTokenValid {
                // Need to refresh
                guard let refreshToken = tokenStore.refreshToken else {
                    print("[Auth] No refresh token available")
                    throw AuthError.tokenExpired
                }

                print("[Auth] Access token expired, refreshing...")
                let response = try await api.refreshTokens(refreshToken)
                tokenStore.accessToken = response.accessToken
                tokenStore.accessTokenExpiry = Date().addingTimeInterval(TimeInterval(response.expiresIn))
                tokenStore.refreshToken = response.refreshToken
                print("[Auth] Token refresh successful")
            }

            // Successfully authenticated
            currentUser = storedUser
            isAuthenticated = true
            authState = .authenticated(user: storedUser)
            print("[Auth] Authentication restored for: \(storedUser.email.prefix(3))***")

            // Notify other services
            NotificationCenter.default.post(name: .userDidAuthenticate, object: nil)

        } catch {
            print("[Auth] Token refresh failed during startup: \(error)")
            print("[Auth] Error type: \(type(of: error)), description: \(error.localizedDescription)")

            if isPermanentAuthError(error) {
                // Permanent error (account blocked, device limit) - force re-login
                print("[Auth] Permanent auth error - user must re-authenticate")
                currentUser = nil
                isAuthenticated = false
                authState = .unauthenticated
            } else {
                // All other errors (network, invalid token, server, decoding, etc.)
                // → Degraded mode: keep user authenticated with cached data.
                // getAccessToken() will retry refresh transparently on next action.
                print("[Auth] Non-permanent error - entering degraded mode with cached user")
                currentUser = storedUser
                isAuthenticated = true
                authState = .authenticated(user: storedUser)
                NotificationCenter.default.post(name: .userDidAuthenticate, object: nil)
            }
        }
    }

    // MARK: - Credentials Login

    /// Login with email and password
    func loginWithCredentials(email: String, password: String) async throws {
        authState = .authenticating

        do {
            let response = try await api.login(
                email: email,
                password: password,
                deviceInfo: deviceInfo
            )

            // Store tokens
            let tokens = AuthTokens(
                accessToken: response.accessToken,
                refreshToken: response.refreshToken,
                expiresIn: response.expiresIn
            )
            tokenStore.storeTokens(tokens, user: response.user)

            // Update state
            currentUser = response.user
            isAuthenticated = true
            authState = .authenticated(user: response.user)

            // Notify other services
            NotificationCenter.default.post(name: .userDidAuthenticate, object: nil)

        } catch {
            authState = .error(message: error.localizedDescription)
            throw error
        }
    }

    // MARK: - Registration

    /// Register a new account with email and password
    func registerWithCredentials(name: String, email: String, password: String) async throws {
        authState = .authenticating

        do {
            let response = try await api.register(
                name: name,
                email: email,
                password: password,
                deviceInfo: deviceInfo
            )

            // Store tokens
            let tokens = AuthTokens(
                accessToken: response.accessToken,
                refreshToken: response.refreshToken,
                expiresIn: response.expiresIn
            )
            tokenStore.storeTokens(tokens, user: response.user)

            // Update state
            currentUser = response.user
            isAuthenticated = true
            authState = .authenticated(user: response.user)

            // Notify other services
            NotificationCenter.default.post(name: .userDidAuthenticate, object: nil)

        } catch {
            authState = .error(message: error.localizedDescription)
            throw error
        }
    }

    // MARK: - Device Code Flow

    /// Start the device code flow for OAuth users
    func startDeviceCodeFlow() async throws -> DeviceCodeResponse {
        authState = .authenticating

        do {
            let response = try await api.requestDeviceCode(deviceInfo: deviceInfo)

            let expiresAt = Date().addingTimeInterval(TimeInterval(response.expiresIn))
            authState = .deviceCodePending(
                userCode: response.userCode,
                deviceCode: response.deviceCode,
                expiresAt: expiresAt
            )

            // Start polling
            startPolling(deviceCode: response.deviceCode, expiresAt: expiresAt)

            return response

        } catch {
            authState = .error(message: error.localizedDescription)
            throw error
        }
    }

    /// Cancel the device code flow
    func cancelDeviceCodeFlow() {
        pollingTask?.cancel()
        pollingTask = nil
        authState = .unauthenticated
    }

    private func startPolling(deviceCode: String, expiresAt: Date) {
        pollingTask?.cancel()

        pollingTask = Task {
            while !Task.isCancelled && Date() < expiresAt {
                do {
                    try await Task.sleep(nanoseconds: UInt64(pollingInterval * 1_000_000_000))

                    if Task.isCancelled { break }

                    let response = try await api.pollDeviceCode(deviceCode)

                    // Check if authorized
                    if let accessToken = response.accessToken,
                       let refreshToken = response.refreshToken,
                       let expiresIn = response.expiresIn,
                       let user = response.user {

                        // Store tokens
                        let tokens = AuthTokens(
                            accessToken: accessToken,
                            refreshToken: refreshToken,
                            expiresIn: expiresIn
                        )

                        await MainActor.run {
                            tokenStore.storeTokens(tokens, user: user)
                            currentUser = user
                            isAuthenticated = true
                            authState = .authenticated(user: user)

                            // Notify other services
                            NotificationCenter.default.post(name: .userDidAuthenticate, object: nil)
                        }

                        return
                    }

                    // Still pending - continue polling
                    if response.error == "authorization_pending" {
                        continue
                    }

                    // Some other error
                    if let error = response.error, error != "authorization_pending" {
                        await MainActor.run {
                            authState = .error(message: response.message ?? error)
                        }
                        return
                    }

                } catch {
                    // Ignore network errors during polling, just continue
                    print("[Auth] Polling error: \(error)")
                }
            }

            // Expired without authorization
            if !Task.isCancelled {
                await MainActor.run {
                    authState = .error(message: "Device code expired. Please try again.")
                }
            }
        }
    }

    // MARK: - Logout

    /// Logout from current device
    func logout() async {
        pollingTask?.cancel()

        do {
            try await api.logout(
                refreshToken: tokenStore.refreshToken,
                allDevices: false
            )
        } catch {
            print("[Auth] Logout API call failed: \(error)")
            // Continue with local logout anyway
        }

        tokenStore.clearAll()
        currentUser = nil
        isAuthenticated = false
        authState = .unauthenticated

        // Notify other services
        NotificationCenter.default.post(name: .userDidLogout, object: nil)
    }

    /// Logout from all devices
    func logoutAllDevices() async {
        pollingTask?.cancel()

        do {
            try await api.logout(
                refreshToken: tokenStore.refreshToken,
                allDevices: true
            )
        } catch {
            print("[Auth] Logout all devices API call failed: \(error)")
        }

        tokenStore.clearAll()
        currentUser = nil
        isAuthenticated = false
        authState = .unauthenticated

        // Notify other services
        NotificationCenter.default.post(name: .userDidLogout, object: nil)
    }

    // MARK: - Token Access

    /// Get a valid access token, refreshing if needed
    func getAccessToken() async throws -> String {
        if tokenStore.isAccessTokenValid, let token = tokenStore.accessToken {
            return token
        }

        guard let refreshToken = tokenStore.refreshToken else {
            throw AuthError.notAuthenticated
        }

        let response = try await api.refreshTokens(refreshToken)
        tokenStore.accessToken = response.accessToken
        tokenStore.accessTokenExpiry = Date().addingTimeInterval(TimeInterval(response.expiresIn))
        tokenStore.refreshToken = response.refreshToken

        guard let token = tokenStore.accessToken else {
            throw AuthError.invalidToken
        }

        return token
    }

    // MARK: - Google Sign-In

    /// Login with Google using ASWebAuthenticationSession
    func loginWithGoogle() async throws {
        authState = .authenticating

        do {
            // Start Google OAuth flow
            let googleResult = try await GoogleAuthService.shared.startGoogleSignIn()

            // Exchange authorization code for tokens with our backend
            let response = try await api.exchangeGoogleAuth(
                code: googleResult.authorizationCode,
                codeVerifier: googleResult.codeVerifier,
                redirectUri: googleResult.redirectUri,
                deviceInfo: deviceInfo
            )

            // Store tokens
            let tokens = AuthTokens(
                accessToken: response.accessToken,
                refreshToken: response.refreshToken,
                expiresIn: response.expiresIn
            )
            tokenStore.storeTokens(tokens, user: response.user)

            // Update state
            currentUser = response.user
            isAuthenticated = true
            authState = .authenticated(user: response.user)

            // Notify other services
            NotificationCenter.default.post(name: .userDidAuthenticate, object: nil)

            print("[Auth] Google login successful, isNewUser: \(response.isNewUser)")

        } catch let error as GoogleAuthError where error == .userCancelled {
            // User cancelled - just return to unauthenticated
            authState = .unauthenticated
            throw error

        } catch {
            authState = .error(message: error.localizedDescription)
            throw error
        }
    }

    // MARK: - Email Check

    /// Check if an email exists and what authentication method it uses
    func checkEmailAuthMethod(_ email: String) async throws -> EmailCheckResponse {
        return try await api.checkEmail(email)
    }

    // MARK: - Error Classification

    /// Errors that indicate the session is gone and the user MUST re-authenticate.
    /// Transient errors (network, server) → degraded mode with transparent retry.
    private func isPermanentAuthError(_ error: Error) -> Bool {
        if let authError = error as? AuthError {
            switch authError {
            case .accountBlocked, .deviceLimitReached:
                // Account-level blocks that won't resolve by retrying
                return true
            case .invalidToken, .tokenExpired:
                // Server explicitly rejected the refresh token (401).
                // The session is gone — retrying will keep failing.
                return true
            default:
                // Network errors, server errors, decoding errors → transient
                // Allow degraded mode with transparent retry via getAccessToken()
                return false
            }
        }

        // All non-AuthError errors (URLError, DecodingError, etc.) → transient
        return false
    }
}
